client/signup/controller.js

```diff
@@ -305,6 +308,38 @@ export default {
 
 		next();
 	},
+	setSelectedSiteForSignup( { store: signupStore, query }, next ) {
+		const { getState, dispatch } = signupStore;
+		const signupDependencies = getSignupDependencyStore( getState() );
+
+		const siteFragment = signupDependencies?.siteSlug || query?.siteSlug;
+		const siteId = getSiteId( getState(), siteFragment );
+		if ( siteId ) {
+			dispatch( setSelectedSiteId( siteId ) );
+			next();
+		} else {
+			// Fetch the site by siteFragment and then try to select again
+			dispatch( requestSite( siteFragment ) ).then( () => {
+				let freshSiteId = getSiteId( getState(), siteFragment );
+
+				if ( ! freshSiteId ) {
+					const wpcomStagingFragment = siteFragment.replace(
+						/\b.wordpress.com/,
+						'.wpcomstaging.com'
+					);
+					freshSiteId = getSiteId( getState(), wpcomStagingFragment );
+				}
+
+				if ( freshSiteId ) {
+					// onSelectedSiteAvailable might render an error page about domain-only sites or redirect
+					// to wp-admin. In that case, don't continue handling the route.
```

Comment:
I'm not quite sure to understand this comment. As far I can see, we call `dispatch()` and `next()` in the exact same way than if we were able to retrieve `siteId` in the first place (above in this method).